---
title: "Cell type annotation"
author: "Workshop on the analysis of scRNA-sequencing data"
format: 
  revealjs: 
    fontsize: 20pt
    slide-number: TRUE
    scrollable: TRUE
    theme: default
    width: 1280
    length: 800
    fig-cap-location: top
    logo: "images/Universitaetsmedizin-eps-converted-to.png"
    footer: "Workshop on the analysis of scRNA-sequencing data"
date: "2024-02-02" 
---

```{r echo = F}
.Last <- function() {          # Beep when knitting ends
  beepr::beep(1)
  Sys.sleep(1)
}

options(error = function(){    # scream on error
  beepr::beep(9)
  Sys.sleep(1)
  }
)
```


## Motivation

Cell type annotation is one of the more complex steps of a single-cell RNA-seq analysis.

However, it is commonly done to support and ease data interpretations.

## Types of Cell type annotation

There are usually two common approaches to cell type annotation:

+ Using a reference data set
+ Using sets of known marker genes for specific cell types

However, both approaches can be flawed and are only as good as the respective reference data used.


```{r, echo=FALSE}
# data used consists of stem cells
```


## Which one is the correct annotation of the data?

::: {layout-ncol="2"}


![](images/TSNE_many.png)


![](images/TSNE_tCells.png)

:::

## Conclusions


+ Cell type annotation can greatly support and facilitate data interpretation.
+ However you should be careful to not view the cell type annotation as a ‘ground truth’
+ In order to really be able to interpret cell type annotation you need to be aware of the cell type composition of your sample and maybe evaluate the results of different reference data sets.

